### ISSUE TITLE
Feature: open data viewer from debug variables view

### ISSUE THREAD

**Issue by amunger** [microsoft/vscode-data-wrangler#131]
There is a new proposed API that would enable the data wrangler extension to contribute its data visualizer for specific variables listed in the debug variables view: https://github.com/microsoft/vscode/blob/b62def4ceb2fe7c0a31908e98e76586d4a191285/src/vscode-dts/vscode.proposed.debugVisualization.d.ts#L1

With a sample implementation here:
https://github.com/microsoft/vscode-extension-samples/tree/connor4312/debug-viz-demo/proposed-api-sample

If that API wouldn't work for some reason, I'm sure @connor4312 would be happy to have your feedback.

**Comment by pwang347** [comment]
Hi @connor4312, it would be great to have this integrated for Data Wrangler.

I wanted to check with you: is the runtime environment exposed for the debugging context or do visualizers only have access to the serialized values?

Data Wrangler works primarily with Pandas DataFrame objects so if it is the latter we would at least need a full representation (e.g. pickled format) to be able to parse it properly. Of course, we would ideally like access to the full runtime environment but from what I'm seeing that doesn't seem to be available.

Thanks!

**Comment by connor4312** [comment]
>  is the runtime environment exposed for the debugging context or do visualizers only have access to the serialized values?

You can see more of a description here https://github.com/microsoft/vscode/issues/197287#issuecomment-1899409209

- The "when" clause in the package.json gives you approximate matching on the variable to avoid having to activate the extension when it's not required
- Then the `DebugVisualizationProvider` is called is called with more details from DAP, as well as the DebugSession, in its context. You can interact with the debugger or do anything you need to here, and set any command to run either in the provide or resolve method

**Comment by pwang347** [comment]
Got it, thanks for the details. I will need to play around with it some more, but tentatively it seems like we can evaluate arbitrary expressions with `setExpression` which is something. We will also need to think more on our side how we want the user experience for this to be.

**Comment by pwang347** [comment]
related: https://github.com/microsoft/vscode-data-wrangler/issues/49

**Comment by amunger** [comment]
After looking into this area more, I don't think the new proposed API is required and you should be able to just contribute to the debug context menu:

```
"debug/variables/context": [
      {
          "command": "jupyter.showDataViewer",
          "group": "1_view",
          "when": "DebugSessionSupportsDataViewer"
      }
  ],
```

Where `DebugSessionSupportsDataViewer` is a context key that you would update by watching `vscode.debug.onActiveDebugSessionChanged`, and checking if that debug session is compatible.

The new API is more targeted towards extensions that want to provide a way to display the variable value within the variable view itself.

**Comment by gkocjan** [comment]
Is there a chance we can implement this feature? Right now, many people are using DataViewer in debug mode from the Jupyter Extension, but it just said that this feature is deprecated in favour of Data Wrangler.

**Comment by pwang347** [comment]
Hi @gkocjan, note that it is possible already to use Data Wrangler while debugging in both the stable and pre-release versions (any of the following entrypoints should work):
![image](https://github.com/microsoft/vscode-data-wrangler/assets/8560030/a139413a-dbe2-4e5e-b109-2d55de736014)


However, specifically for this button in the debug panel

<img width="242" alt="image" src="https://github.com/microsoft/vscode-data-wrangler/assets/8560030/1ce515cc-5bac-479a-a3fb-c159418c12e0">

... it's something that still needs to be hooked up. You can track the progress on it here: https://github.com/microsoft/vscode-jupyter/issues/15680

Thanks!

**Comment by juanbretti** [comment]
Hello @pwang347 .
What you are proposing, is while using iPython (interactive Python), Jupyter Notebook or interactive Window.

# Standard Python Script
When executing a standard Python Script (.py), this option is not available.
For example, when debugging this file:

![image](https://github.com/microsoft/vscode-data-wrangler/assets/10165911/2adec314-de4e-41e0-aa70-9be0e8e2e66c)

There is only this option:

![image](https://github.com/microsoft/vscode-data-wrangler/assets/10165911/57e68b7b-71b3-470e-9020-821d3927064d)

And no other variables inspector.

# Using interactive Python

When using interactive Python interpreter, there is the option:

![image](https://github.com/microsoft/vscode-data-wrangler/assets/10165911/87c7dfdc-82cb-456d-a306-48c150b6f9f3)

Here is the action when clicking the "Show variable snapshot in data view":

![image](https://github.com/microsoft/vscode-data-wrangler/assets/10165911/0a28fd29-d3fd-47e9-8ad2-9cd93cc3bd17)

# What is missing
What is missing is the ability to use the **Data Wrangler** when debugging a **Python Script**, while using the standard compiler.
What do you think?

# Code used:
```python
# %%
import pandas as pd

# Creating a dictionary with sample data
data = {
    'Name': ['John', 'Anna', 'Peter', 'Linda', 'Mark'],
    'Age': [25, 30, 35, 28, 40],
    'City': ['New York', 'Paris', 'London', 'Tokyo', 'Sydney'],
    'Salary': [50000, 60000, 70000, 55000, 65000]
}

# Creating DataFrame
df = pd.DataFrame(data)

# %%
print(df)

# %%
```

**Comment by pwang347** [comment]
Hi @juanbretti, thank you for sharing more about your use case!

Just to confirm, I found that the "View Value in Data Viewer" button only appears if you have the Jupyter extension installed but it isn't present otherwise (e.g. if you disable the Jupyter extension). Let me know if my steps below are the same as yours:

**Starting the debugger**
<img width="297" alt="image" src="https://github.com/microsoft/vscode-data-wrangler/assets/8560030/25fa50e7-b489-40bf-94ab-c48a1ddff846">

**With extension**
<img width="502" alt="image" src="https://github.com/microsoft/vscode-data-wrangler/assets/8560030/63c28cd5-c182-4f9c-b3d0-807809432ff3">

**Without**
<img width="925" alt="image" src="https://github.com/microsoft/vscode-data-wrangler/assets/8560030/9836e722-052c-4e37-9ace-7b5b72cf274c">

If having the Jupyter extension installed is not a problem for you, then this issue should address your ask: https://github.com/microsoft/vscode-jupyter/issues/15680

Otherwise, if you want to support this functionality for pure Python without any IPython dependency it would be good to open a new GitHub issue and continue the discussion there. Thanks!

**Comment by juanbretti** [comment]
@pwang347 , thank you for your prompt response.
I do have installed the extensions:
- Jupyter
- Data Wrangler
- Python
- Python Debugger

When I debug using:

![image](https://github.com/microsoft/vscode-data-wrangler/assets/10165911/dcaa6c5d-44d3-49e8-a458-63b934e5a692)

I do see:

![image](https://github.com/microsoft/vscode-data-wrangler/assets/10165911/4692beb9-100e-412e-810a-4cbd954b273d)

But this option, opens:

![image](https://github.com/microsoft/vscode-data-wrangler/assets/10165911/86c74698-6a4f-4271-bf40-9a19d25ab48a)

Which is not the "Data Wrangler" variable inspector.

Does it work the same way for you too?

**Comment by pwang347** [comment]
> @pwang347 , thank you for your prompt response. I do have installed the extensions:
> 
> * Jupyter
> * Data Wrangler
> * Python
> * Python Debugger
> 
> When I debug using:
> 
> ![image](https://private-user-images.githubusercontent.com/10165911/329343660-dcaa6c5d-44d3-49e8-a458-63b934e5a692.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyNzkwOTEsIm5iZiI6MTcxNTI3ODc5MSwicGF0aCI6Ii8xMDE2NTkxMS8zMjkzNDM2NjAtZGNhYTZjNWQtNDRkMy00OWU4LWE0NTgtNjNiOTM0ZTVhNjkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA5VDE4MTk1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkwMzYxOWY4MmE4N2Q4NTUzMWNkMzkwYjljNTY0NjJlOWE0ZDgyYWM4OWYyMmQ5YWZmZTU4OTBlMTlhYWYxODAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OJt0OkHVyoLQd721HLZb4nc7eahVsXtJXUoaDvgMzpc)
> 
> I do see:
> 
> ![image](https://private-user-images.githubusercontent.com/10165911/329343886-4692beb9-100e-412e-810a-4cbd954b273d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyNzkwOTEsIm5iZiI6MTcxNTI3ODc5MSwicGF0aCI6Ii8xMDE2NTkxMS8zMjkzNDM4ODYtNDY5MmJlYjktMTAwZS00MTJlLTgxMGEtNGNiZDk1NGIyNzNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA5VDE4MTk1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2NjBlYWQ3ZWU5OGFlMjQwMzViODlhYjAwODNkOWVkMTBkZWQwNDk2YzBkY2E5MTViMmFhYTgzY2I4NmU1MDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OYSaH-A9-BgIt0qPejz1yAsvUt_Gesdy39a2CzunFc4)
> 
> But this option, opens:
> 
> ![image](https://private-user-images.githubusercontent.com/10165911/329344427-86c74698-6a4f-4271-bf40-9a19d25ab48a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyNzkwOTEsIm5iZiI6MTcxNTI3ODc5MSwicGF0aCI6Ii8xMDE2NTkxMS8zMjkzNDQ0MjctODZjNzQ2OTgtNmE0Zi00MjcxLWJmNDAtOWExOWQyNWFiNDhhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA5VDE4MTk1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkZWRiMTgwYTZlNDQ5ZWRhZWE4OGJiZTYwNDgyN2MwOTk0NGYxNmZhODhiZjhiNTE2ZTA5M2VjYzhhOTI5N2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uDbdYb_TY7QBVoHsP9lBxLWsmCOorj8JDEu3oQ5e_OU)
> 
> Which is not the "Data Wrangler" variable inspector.
> 
> Does it work the same way for you too?

Hi Juan, yes that is the way it works for me as well. I meant to say that the viewer you see is coming from the Jupyter extension currently, so there is technically a dependency on IPython (since it requires having a Python kernel). If that is not an issue to you, the missing functionality you described should be resolved soon from the Jupyter side. Hope this helps!

**Comment by 7Eric9** [comment]
can't wait to use this feature!!!

**Comment by gkocjan** [comment]
https://github.com/microsoft/vscode-data-wrangler/issues/131#issuecomment-2099977714
My point was that the Jupyter extension showed a message that DataViewer is deprecated in favour of DataWrangler, and there is the missing feature of debugging pure .py scripts that was really well shown by @juanbretti here:  https://github.com/microsoft/vscode-data-wrangler/issues/131#issuecomment-2103109736

TLDR: it would be great to use Data Wrangler from a pure Python script debugging window because it gives many more features

**Comment by 7Eric9** [comment]
@pwang347 I didn't find the entry for .py, did I missed sth? like configuration etc...
![image](https://github.com/microsoft/vscode-data-wrangler/assets/3244976/fa6afb61-fdab-4922-98f3-3b058e368872)

**Comment by juanbretti** [comment]
> Maybe I should switch my vsocde from stable to insider?

Yes, as mentioned before by @pwang347 , you need to try [Visual Studio Code Insiders](https://code.visualstudio.com/insiders/).

**Comment by pwang347** [comment]
Hi @7Eric9! Yes, that is correct. If you would like to try out the experience now, you will need to use VS Code Insiders.

For a bit of explanation, this is because the Jupyter prerelease version needed above (v2024.5.2024052401) depends on the latest VS Code Insiders version, which is not yet available in VS Code Stable (but will be in about a week).

**Comment by edumoros** [comment]
Hello @pwang347, do you know when this feature will be available in VS Code Stable? You mentioned that it should be in early June, but I haven't seen it in the last version

**Comment by pwang347** [comment]
> Hello @pwang347, do you know when this feature will be available in VS Code Stable? You mentioned that it should be in early June, but I haven't seen it in the last version

Hi @edumoros! It is working for me on the following latest stable versions:
* VS Code Stable: 1.91.0
* Data Wrangler Extension: 1.4.1
* Jupyter Extension: 2024.6.0

Could you please double-check if the same configuration above works for you?

**Comment by edumoros** [comment]
Hi @pwang347! I have those same versions, but it's not working for me.

![Screenshot 2024-07-08 at 20 38 31](https://github.com/microsoft/vscode-data-wrangler/assets/93947494/00173f5e-7dbe-470d-8550-b76bf05f64a6)

**Comment by pwang347** [comment]
@edumoros That's odd, could you please double check that the Data Wrangler extension is not disabled?

If Data Wrangler is not disabled, I would recommend opening an issue here: https://github.com/microsoft/vscode-jupyter/issues

Since the Data Viewer is being routed by the Jupyter extension and you should not see the old one if you have Data Wrangler installed. Thanks!

**Comment by pwang347** [comment]
@Huanyu-Liu Hi! The launch delegation is currently happening on the Jupyter extension side, could you please open an issue in the [Jupyter extension repository](https://github.com/microsoft/vscode-jupyter/issues)?

It would be great if you can provide short example code to reproduce the issue. Thanks!

**Comment by TheAjaykrishnanR** [comment]
[Data Wrangler](https://code.visualstudio.com/docs/datascience/data-wrangler) for pytorch makes it very easy to quickly compare values while debugging. printing tensors with long dimensions onto the console buffer is very inefficient unless you splice them.
